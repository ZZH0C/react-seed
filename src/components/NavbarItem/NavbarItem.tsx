import React from 'react';
import classNames from 'classnames';

interface NavbarElemProps extends React.HTMLAttributes<HTMLElement> {
  isActive: boolean;
  url?: string;
}

export const NavbarItem: React.FC<NavbarElemProps> = ({
  isActive,
  url,
  children,
}) => {
  return (
    <li>
      <a
        href={url}
        title="Menu item"
        className={classNames('', { active: isActive })}
      >
        {children}
      </a>
    </li>
  );
};
