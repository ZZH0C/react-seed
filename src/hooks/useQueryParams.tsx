import queryString from 'query-string';
import { useLocation } from 'react-router-dom';
import _ from 'lodash';
import { useCallback } from 'react';

interface queryProps {
  category: string;
  key: 'in' | 'category' | 'search';
}

export const useQueryParams = (): {
  getGoogleQueryParamsCallback: () => string;
  changeParamsCallback: (
    category: queryProps['category'],
    key: queryProps['key'],
  ) => { isActive: boolean; parsedParams: string };
} => {
  const location = useLocation();
  const getGoogleQueryParams = () => {
    const queryParams = queryString.parse(location.search);
    if (!queryParams.label && !queryParams.category && !queryParams.search)
      return '';
    let result = '';
    _.forEach(queryParams, (value, key) => {
      if (key === 'search') {
        result += ` ${value} `;
      } else {
        result += `${key}:${value} `;
      }
    });

    return _.trim(result);
  };
  const changeParams = (
    category: queryProps['category'],
    key: queryProps['key'],
  ): { isActive: boolean; parsedParams: string } => {
    let isActive = false;
    const queryParams = queryString.parse(location.search);
    if (queryParams[key] === category) {
      isActive = true;
    }
    queryParams[key] = category;
    const parsedParams = queryString.stringify(queryParams);
    return {
      isActive: isActive,
      parsedParams: parsedParams,
    };
  };

  const changeParamsCallback = useCallback(
    (category: string, key: queryProps['key']) => {
      return changeParams(category, key);
    },
    [changeParams],
  );
  const getGoogleQueryParamsCallback = useCallback(() => {
    return getGoogleQueryParams();
  }, [getGoogleQueryParams]);

  return {
    changeParamsCallback,
    getGoogleQueryParamsCallback,
  };
};
