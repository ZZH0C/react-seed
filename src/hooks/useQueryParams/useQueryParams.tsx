import queryString from 'query-string';
import { useLocation } from 'react-router-dom';
import _ from 'lodash';
import { useCallback, useMemo } from 'react';

interface queryProps {
  category: string;
  key: 'in' | 'category' | 'search';
}

export const useQueryParams = (): {
  getGoogleQueryParams: () => string;
  changeParams: (
    category: queryProps['category'],
    key: queryProps['key'],
  ) => { isActive: boolean; parsedParams: string };
} => {
  const location = useLocation();

  const changeParams = useCallback(
    (category: queryProps['category'], key: queryProps['key']) => {
      let isActive = false;
      const queryParams = queryString.parse(location.search);
      if (queryParams[key] === category) {
        isActive = true;
      }
      queryParams[key] = category;
      const parsedParams = queryString.stringify(queryParams);
      return {
        isActive: isActive,
        parsedParams: parsedParams,
      };
    },
    [location.search],
  );
  const getGoogleQueryParams = useCallback(() => {
    const queryParams = queryString.parse(location.search);
    if (!queryParams.label && !queryParams.category && !queryParams.search)
      return '';
    let result = '';
    _.forEach(queryParams, (value, key) => {
      if (key === 'search') {
        result += ` ${value} `;
      } else {
        result += `${key}:${value} `;
      }
    });

    return _.trim(result);
  }, [location.search]);

  return useMemo(
    () => ({
      changeParams,
      getGoogleQueryParams,
    }),
    [changeParams, getGoogleQueryParams],
  );
};
