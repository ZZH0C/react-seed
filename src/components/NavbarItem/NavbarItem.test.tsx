import { configure, shallow } from 'enzyme';
import { NavbarItem } from './NavbarItem';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
import { useQueryParams } from '../../hooks/useQueryParams/useQueryParams';
import { mocked } from 'ts-jest/utils';

jest.mock('../../hooks/useQueryParams/useQueryParams');
const mockedQueryParams = mocked(useQueryParams).mockImplementation(() => {
  const changeParams = () => {
    return {
      parsedParams: 'someText',
      isActive: true,
    };
  };
  const getGoogleQueryParams = () => {
    return 'someText';
  };
  return { getGoogleQueryParams, changeParams };
});

describe('components/NavbarItem', () => {
  const mockProps = {
    href: '/',
    label: 'Label',
    name: 'name',
    children: 'some text',
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render', () => {
    expect(
      shallow(
        <NavbarItem
          href={mockProps.href}
          label={mockProps.label}
          name={mockProps.name}
        >
          {mockProps.children}
        </NavbarItem>,
      ),
    ).toMatchSnapshot();
  });

  it('should call changeParams', () => {
    shallow(
      <NavbarItem
        href={mockProps.href}
        label={mockProps.label}
        name={mockProps.name}
      >
        {mockProps.children}
      </NavbarItem>,
    );
    expect(mockedQueryParams).toHaveBeenCalledTimes(1);
  });
});
