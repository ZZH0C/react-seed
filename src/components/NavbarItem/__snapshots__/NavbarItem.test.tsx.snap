// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/NavbarItem should match snapshot 1`] = `
<li>
  <Link
    className="active"
    title="name"
    to={
      Object {
        "pathname": "/",
        "search": "someText",
      }
    }
  >
    some text
  </Link>
</li>
`;

exports[`components/NavbarItem should render 1`] = `
<li>
  <Link
    className="active"
    title="name"
    to={
      Object {
        "pathname": "/",
        "search": "someText",
      }
    }
  >
    some text
  </Link>
</li>
`;
